Source/WebCore/Modules/speech/SpeechSynthesisUtterance.cpp

```diff
@@ -81,26 +81,49 @@ void SpeechSynthesisUtterance::setVoice(SpeechSynthesisVoice* voice)
 
 void SpeechSynthesisUtterance::eventOccurred(const AtomString& type, unsigned long charIndex, unsigned long charLength, const String& name)
 {
+    ASSERT(m_isActive);
+
     if (m_completionHandler) {
-        if (type == eventNames().endEvent)
+        if (type == eventNames().endEvent) {
             m_completionHandler(*this);
+            setIsActive(false);
+        }
 
         return;
     }
 
     dispatchEvent(SpeechSynthesisEvent::create(type, { this, charIndex, charLength, static_cast<float>((MonotonicTime::now() - startTime()).seconds()), name }));
+    if (type == eventNames().endEvent)
+        setIsActive(false);
 }
 
 void SpeechSynthesisUtterance::errorEventOccurred(const AtomString& type, SpeechSynthesisErrorCode errorCode)
 {
+    ASSERT(m_isActive);
+
     if (m_completionHandler) {
         m_completionHandler(*this);
         return;
     }
 
     dispatchEvent(SpeechSynthesisErrorEvent::create(type, { { this, 0, 0, static_cast<float>((MonotonicTime::now() - startTime()).seconds()), { } }, errorCode }));
+    setIsActive(false);
+}
+
+void SpeechSynthesisUtterance::setIsActive(bool isActive)
```

Comment:
Can we use activity token instead of this boolean? (ActiveDOMObject::makePendingActivity)

Reply:
Do you mean to make a `PendingActivity` on `setIsActive(true)` and clear it when `setIsActive(false)`? 
Then how do we check event listeners (and their changes)? Or do you mean we should keep object alive even when there's no event listener?

Reply:
I don't think we need to check whether there is an event listener or not. Keeping it alive whenever there is a pending event is okay.

Reply:
Ok, updated patch based on comment